### PR TITLE
icu: fix cross-build to iOS/tvOS/watchOS with apple-clang

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

fixes https://github.com/conan-io/conan-center-index/issues/6648

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/icu/all/conanfile.py [hunk 1]
```diff
@@ -134,19 +138,23 @@ def _workaround_icu_20545(self):
...
+            ("Linux", "clang"): "Linux",
+            ("Macos", "gcc"): "MacOSX",
+            ("Macos", "clang"): "MacOSX",
+            ("Macos", "apple-clang"): "MacOSX",
+            ("iOS", "apple-clang"): "MacOSX",
+            ("tvOS", "apple-clang"): "MacOSX",
+            ("watchOS", "apple-clang"): "MacOSX",
+            ("FreeBSD", "gcc"): "FreeBSD",
+            ("FreeBSD", "clang"): "FreeBSD",
+        }.get((str(self.settings.os), str(self.settings.compiler)))
```
Comment (SSE4) [inline recipes/icu/all/conanfile.py:157]: we need to improve an error, might do in this PR or another PR, as you wish
something like:
1. extracting `platform` to a property
2. in validate add some code like:
```
def validate(self):
    if self.settings.os is not in self._platform.keys():
        raise ConanInvalidConfiguration("unsupported platform, give up")
```

Comment (SpaceIm) [reply recipes/icu/all/conanfile.py:157]: I agree

Comment (SpaceIm) [reply recipes/icu/all/conanfile.py]: done

**Review by madebr — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]
ICU - I see you!

**Comment by SpaceIm** [comment]
I forgot something in package_info(): icutu & icutest are not created when tools are disabled.

**Review by madebr — APPROVED** [review]

File: recipes/icu/all/conanfile.py [hunk 2]
```diff
@@ -61,15 +66,45 @@ def configure(self):
...
+            raise ConanInvalidConfiguration(
+                "Compiling ICU for {} with {} not supported yet".format(str(self.settings.os),
+                                                                        str(self.settings.compiler)))
+
+    @property
+    def _platform(self):
+        return {
+            ("Windows", "Visual Studio"): "Cygwin/MSVC",
+            ("Windows", "gcc"): "MinGW",
+            ("AIX", "gcc"): "AIX/GCC",
```
Comment (madebr) [inline recipes/icu/all/conanfile.py:80]: I don't know whether this is in-scope, but this is missing a line for Android.
https://github.com/unicode-org/icu/blob/b03b8be7414422d8198b77da4c7ed8c9457dd349/icu4c/source/runConfigureICU#L48
```suggestion
            ("Android", "clang"): "Linux",
            ("AIX", "gcc"): "AIX/GCC",
```

File: recipes/icu/all/conanfile.py [hunk 3]
```diff
@@ -134,22 +169,9 @@ def _workaround_icu_20545(self):
...
-                    ("Macos", "gcc"): "MacOSX",
-                    ("Macos", "clang"): "MacOSX",
-                    ("Macos", "apple-clang"): "MacOSX",
-                    ("FreeBSD", "gcc"): "FreeBSD",
-                    ("FreeBSD", "clang"): "FreeBSD"}.get((str(self.settings.os),
-                                                          str(self.settings.compiler)))
         arch64 = ['x86_64', 'sparcv9', 'ppc64', 'ppc64le', 'armv8', 'armv8.3', 'mips64']
         bits = "64" if self.settings.arch in arch64 else "32"
-        args = [platform,
+        args = [self._platform,
```
Comment (madebr) [inline recipes/icu/all/conanfile.py:174]: Looks nicer :P
```suggestion
        args = [
            self._platform,
```

File: recipes/icu/all/conanfile.py [hunk 4]
```diff
@@ -61,15 +66,45 @@ def configure(self):
...
+            ("Linux", "clang"): "Linux",
+            ("Macos", "gcc"): "MacOSX",
+            ("Macos", "clang"): "MacOSX",
+            ("Macos", "apple-clang"): "MacOSX",
+            ("iOS", "apple-clang"): "MacOSX",
+            ("tvOS", "apple-clang"): "MacOSX",
+            ("watchOS", "apple-clang"): "MacOSX",
+            ("FreeBSD", "gcc"): "FreeBSD",
+            ("FreeBSD", "clang"): "FreeBSD",
+        }.get((str(self.settings.os), str(self.settings.compiler)))
```
Comment (madebr) [inline recipes/icu/all/conanfile.py:93]: It should be an error when platform is not available (`None`).
Perhaps test this in `validate`?
Right now, `None` is added to the argument list which fails when doing `" ".join(args)`

Comment (SpaceIm) [reply recipes/icu/all/conanfile.py:93]: It's tested in validate(). It has been added in https://github.com/conan-io/conan-center-index/pull/6652/commits/656a11d13334357211b9f0f70aaacb57f1384a1c

Comment (madebr) [reply recipes/icu/all/conanfile.py:93]: Sorry, I didn't see. I was looking at the current recipe in master.

**Review by SSE4 — APPROVED** [review]

File: recipes/icu/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/icu/all/conanfile.py:80]: Yes it's not in scope of this PR, and I've not tested Android.

Comment (madebr) [reply recipes/icu/all/conanfile.py:80]: I just tried it with the icu recipe on current master, and it's working.
```patch
--- a/recipes/icu/all/conanfile.py
+++ b/recipes/icu/all/conanfile.py
@@ -136,6 +136,7 @@ class ICUBase(ConanFile):
         prefix = self.package_folder.replace("\\", "/")
         platform = {("Windows", "Visual Studio"): "Cygwin/MSVC",
                     ("Windows", "gcc"): "MinGW",
+                    ("Android", "clang"): "Linux",
                     ("AIX", "gcc"): "AIX/GCC",
                     ("AIX", "xlc"): "AIX",
                     ("SunOS", "gcc"): "Solaris/GCC",
```

When I run `conan create . icu/69.1@ -pr:b default64 -pr:h android22b`, icu is build twice: once for x86_64 and once for android.
Do you know how I can avoid rebuilding the x86_64 package when running this command?
For speed purposes.

Comment (madebr) [reply recipes/icu/all/conanfile.py:80]: `-b missing` will skip building everything and just run the test package.

Comment (SpaceIm) [reply recipes/icu/all/conanfile.py:80]: Unfortunately, there are no `-b:h` and `-b:b`.

Comment (madebr) [reply recipes/icu/all/conanfile.py:80]: Would be nice to have that as a time saver.

**Review by prince-chrismc — APPROVED** [review]

File: recipes/icu/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/icu/all/conanfile.py:80]: You may also have to override host triplet if Android like I did if iOS/tvOS/watchOS.
